File: CUTEProject/bundles/ch.hsr.ifs.testframework/src/ch/hsr/ifs/testframework/model/TestSession.java
/*******************************************************************************
 * Copyright (c) 2007-2011, IFS Institute for Software, HSR Rapperswil,
 * Switzerland, http://ifs.hsr.ch
 *
 * Permission to use, copy, and/or distribute this software for any
 * purpose without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 ******************************************************************************/
package ch.hsr.ifs.testframework.model;

import java.util.ArrayList;
import java.util.List;

import org.eclipse.debug.core.ILaunch;


public class TestSession implements ITestComposite {

   private final List<TestElement>            rootElements = new ArrayList<>();
   private final List<ITestCompositeListener> listeners    = new ArrayList<>();;

   private final ILaunch launch;

   public TestSession(ILaunch launch) {
      super();
      this.launch = launch;
   }

   public List<TestElement> getRootElements() {
      return rootElements;
   }

   public ILaunch getLaunch() {
      return launch;
   }

   @Override
   public void addTestElement(TestElement element) {
      rootElements.add(element);
      element.setParent(this);
      for (ITestCompositeListener lis : listeners) {
         lis.newTestElement(this, element);
      }
   }

   @Override
   public List<TestElement> getElements() {
      return rootElements;
   }

   @Override
   public int getError() {
      int tot = 0;
      for (TestElement tElement : rootElements) {
         if (tElement instanceof ITestComposite) {
            ITestComposite testComp = (ITestComposite) tElement;
            tot += testComp.getError();
         } else if (tElement instanceof TestCase) {
            TestCase tCase = (TestCase) tElement;
            if (tCase.getStatus() == TestStatus.error) {
               ++tot;
            }
         }
      }
      return tot;
   }

   @Override
   public int getFailure() {
      int tot = 0;
      for (TestElement tElement : rootElements) {
         if (tElement instanceof ITestComposite) {
            ITestComposite testComp = (ITestComposite) tElement;
            tot += testComp.getFailure();
         } else if (tElement instanceof TestCase) {
            TestCase tCase = (TestCase) tElement;
            if (tCase.getStatus() == TestStatus.failure) {
               ++tot;
            }
         }
      }
      return tot;
   }

   @Override
   public int getRun() {
      int tot = 0;
      for (TestElement tElement : rootElements) {
         if (tElement instanceof ITestComposite) {
            ITestComposite testComp = (ITestComposite) tElement;
            tot += testComp.getRun();
         } else if (tElement instanceof TestCase) {
            TestCase tCase = (TestCase) tElement;
            if (tCase.getStatus() == TestStatus.error || tCase.getStatus() == TestStatus.failure || tCase.getStatus() == TestStatus.success) {
               ++tot;
            }
         }
      }
      return tot;
   }

   @Override
   public int getSuccess() {
      int tot = 0;
      for (TestElement tElement : rootElements) {
         if (tElement instanceof ITestComposite) {
            ITestComposite testComp = (ITestComposite) tElement;
            tot += testComp.getSuccess();
         } else if (tElement instanceof TestCase) {
            TestCase tCase = (TestCase) tElement;
            if (tCase.getStatus() == TestStatus.success) {
               ++tot;
            }
         }
      }
      return tot;
   }

   @Override
   public int getTotalTests() {
      int tot = 0;
      for (TestElement tElement : rootElements) {
         if (tElement instanceof ITestComposite) {
            ITestComposite testComp = (ITestComposite) tElement;
            tot += testComp.getTotalTests();
         } else if (tElement instanceof TestCase) {
            ++tot;
         }
      }
      return tot;
   }

   @Override
   public boolean hasErrorOrFailure() {
      return getFailure() + getError() > 0;
   }

   @Override
   public void addListener(ITestCompositeListener listener) {
      if (!listeners.contains(listener)) {
         listeners.add(listener);
      }
   }

   @Override
   public void removeListener(ITestCompositeListener listener) {
      listeners.remove(listener);
   }

   @Override
   public String getRerunName() {
      // errr? what? primitive obsession? (lfelber)
      return ""; // empty means all tests
   }

}

File: CUTEProject/bundles/ch.hsr.ifs.testframework/src/ch/hsr/ifs/testframework/model/TestSession.kt
/*******************************************************************************
 * Copyright (c) 2007-2011, IFS Institute for Software, HSR Rapperswil,
 * Switzerland, http://ifs.hsr.ch
 *
 * Permission to use, copy, and/or distribute this software for any
 * purpose without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 ******************************************************************************/
package ch.hsr.ifs.testframework.model;

import org.eclipse.debug.core.ILaunch;

private fun Boolean.toInt() = if (this) 1 else 0

class TestSession(val launch: ILaunch?) : ITestComposite {

   private val listeners = mutableListOf<ITestCompositeListener>()

   public val rootElements = mutableListOf<TestElement>()

   override fun addTestElement(element: TestElement) {
      rootElements.add(element);
      element.setParent(this);
      listeners.forEach{ it.newTestElement(this, element) }
   }

   override fun getElements() = rootElements


   override fun getError() = 
      rootElements.fold(0, {acc, te ->
         when(te) {
            is ITestComposite -> acc + te.getError()
            is TestCase -> acc + (te.getStatus() == TestStatus.error ).toInt()
            else -> acc
         }
      })

   override fun getFailure() =
      rootElements.fold(0, {acc, te ->
         when(te) {
            is ITestComposite -> acc + te.getFailure()
            is TestCase -> acc + (te.getStatus() == TestStatus.failure).toInt()
            else -> acc
         }
      })

   override fun getRun() =
      rootElements.fold(0, {acc, te ->
         when(te) {
            is ITestComposite -> acc + te.getRun()
            is TestCase -> acc + when(te.getStatus()){
               TestStatus.error, TestStatus.failure, TestStatus.success -> 1
               else -> 0
            }
            else -> acc
         }
      })

   override fun getSuccess() =
      rootElements.fold(0, {acc, te ->
         when(te) {
            is ITestComposite -> acc + te.getSuccess()
            is TestCase -> acc + (te.getStatus() == TestStatus.success).toInt()
            else -> acc
         }
      })

   override fun getTotalTests() =
      rootElements.fold(0, {acc, te ->
         when(te) {
            is ITestComposite -> acc + te.getTotalTests()
            is TestCase ->  acc + 1
            else -> acc
         }
      })

   override fun hasErrorOrFailure() = getFailure() + getError() > 0

   override fun addListener(listener: ITestCompositeListener) {
      if (!listeners.contains(listener)) {
         listeners.add(listener);
      }
   }

   override fun removeListener(listener: ITestCompositeListener) {
      listeners.remove(listener);
   }

   override fun getRerunName() = ""

}
